src/primitives/iter.rs

```diff
@@ -0,0 +1,521 @@
+//! let iterator = witness_prog
+//!     .iter()
+//!     .copied()
+//!     .bytes_to_fes()
+//!     .hrpstring_chars::<Bech32>(&hrp, witness_version);
+//! let hrpstring: String = iterator.collect();
+//! assert_eq!(hrpstring.to_uppercase(), "BC1QW508D6QEJXTDG4Y5R3ZARVARY0C5XW7KV8F3T4");
+//!
+//! // Or if you just want to iterate over the field-elements and the checksum.
+//! let iterator = witness_prog
+//!     .iter()
+//!     .copied()
+//!     .bytes_to_fes()
+//!     .hrp_checksummed::<Bech32>(&hrp, Some(witness_version));
+//! ```
+
+use crate::primitives::checksum::{self, Checksum, PackedFe32};
+use crate::primitives::gf32::Fe32;
+use crate::primitives::hrp::{self, Hrp};
+
+/// Extension trait for byte iterators which provides an adaptor to GF32 elements.
+pub trait ByteIterExt: Sized + Iterator<Item = u8> {
+    /// Adapts the byte iterator to output GF32 field elements instead.
+    ///
+    /// If the total number of bits is not a multiple of 5 we pad with 0s
+    fn bytes_to_fes(mut self) -> BytesToFes<Self> {
+        BytesToFes { last_byte: self.next(), bit_offset: 0, iter: self }
+    }
+}
+
+impl<I> ByteIterExt for I where I: Iterator<Item = u8> {}
+
+/// Extension trait for field element iterators.
+pub trait Fe32IterExt: Sized + Iterator<Item = Fe32> {
+    /// Adapts the `Fe32` iterator to output bytes instead.
+    ///
+    /// If the total number of bits is not a multiple of 8, any trailing bits
+    /// are simply dropped.
+    fn fes_to_bytes(mut self) -> FesToBytes<Self> {
+        FesToBytes { last_fe: self.next(), bit_offset: 0, iter: self }
+    }
+
+    /// Adapts the `Fe32` data iterator to yield characters of the bech32 encode string.
+    ///
+    /// This means inputting the `hrp` to the checksum algorithm, yielding the HRP (and separator),
+    /// prepending the `witness_version` if preset, and then appending a checksum - all as chars.
+    fn hrpstring_chars<Ck: Checksum>(
+        self,
+        hrp: &Hrp,
+        witness_version: Fe32,
+    ) -> HrpstringChars<Self, Ck> {
+        self.checksummed::<Ck>()
+            .with_witness_version(witness_version)
+            .hrp_checksummed(hrp)
+            .hrpstring_chars()
+    }
+
+    /// Adapts the `Fe32` data iterator using by inputting the `hrp`, prepending the `witness_version`
+    /// if preset, and then appends a checksum.
+    fn hrp_checksummed<Ck: Checksum>(
+        self,
+        hrp: &Hrp,
+        witness_version: Option<Fe32>,
+    ) -> HrpChecksummed<Self, Ck> {
+        match witness_version {
+            Some(witness_version) =>
+                Checksummed::new(self).with_witness_version(witness_version).hrp_checksummed(hrp),
+            None => Checksummed::new(self).hrp_checksummed(hrp),
+        }
+    }
+
+    /// Adapts the Fe32 iterator to append a checksum to the end of the data.
+    fn checksummed<Ck: Checksum>(self) -> Checksummed<Self, Ck> { Checksummed::new(self) }
```

Comment:
In caa79f50d9d74b99775dfcc14f66c4140e20397c:

It feels a little weird to have `hrp_checksummed` and also `checksummed`. It seems like maybe the HRP one should only prepend an HRP and not do anything else, so if you want both, you can do `.hrp(hrp, witness_version).checksummed()` which is the same number of characters but more flexible.

Though `.hrp` should probably be `.prepend_hrp`.

And I'd argue we should separate out the witness version and add a `.prepend_witness_version`.

Reply:
Cheers, I'll have another go.